### PR TITLE
[SPARK-50675][SQL] Table and view level collations support

### PR DESCRIPTION
**Pull request by dejankrak-db**
### What changes were proposed in this pull request?

This change introduces table and view level collations support in Spark SQL, allowing CREATE TABLE, ALTER TABLE and CREATE VIEW commands to specify DEFAULT COLLATION to be used. For CREATE commands, this refers to all the underlying columns added as part of the table/view creation. For ALTER TABLE command, this refers to only newly created columns in the future, whereas existing ones are not affected, i.e. their collation remains the same.

The PR has been modelled after the original changes made by @stefankandic in https://github.com/apache/spark/pull/48090, with this PR covering table and view-level collations, whereas a follow up PR will be made covering schema-level collations.

This PR is adding/extending the corresponding DDL commands for specifying table/view level collation, whereas a follow up PR will be created separately to leverage the table/view collation in order to determine default collations for input queries of DML commands.

### Why are the changes needed?

From our internal users feedback, many people would like to be able to specify collation for their objects, instead of each individual columns. This change adds support for table and view level collations, whereas subsequent changes will add support for other objects such as schema-level collations.

### Does this PR introduce _any_ user-facing change?

The change follows the agreed additions in syntax for collation support.

The following syntax is now supported (**bold** parts denote additions):

{ { [CREATE OR] REPLACE TABLE | CREATE [EXTERNAL] TABLE [ IF NOT EXISTS ] }
  table_name
  [ table_specification ]
  [ USING data_source ]
  [ table_clauses ]
  [ AS query ] }

table_specification
  ( { column_identifier column_type [ column_properties ] ] } [, ...]
    [ , table_constraint ] [...] )

table_clauses
  { OPTIONS clause |
    PARTITIONED BY clause |
    CLUSTER BY clause |
    clustered_by_clause |
    LOCATION path [ WITH ( CREDENTIAL credential_name ) ] |
    COMMENT table_comment |
    TBLPROPERTIES clause |
    **DEFAULT COLLATION table_collation_name |**
    WITH { ROW FILTER clause } } [...]

CREATE [ OR REPLACE ] [ TEMPORARY ] VIEW [ IF NOT EXISTS ] view_name
    [ column_list ]
    [ schema_binding |
      COMMENT view_comment |
      TBLPROPERTIES clause |
      **DEFAULT COLLATION collation_name** ] [...]
    AS query

ALTER TABLE table_name
   { ADD COLUMN clause |
     ALTER COLUMN clause |
     DROP COLUMN clause |
     RENAME COLUMN clause |
     **DEFAULT COLLATION clause | …**
   }

### How was this patch tested?

Tests for the new syntax/functionality were added as part of the change. Also, some of the existing tests were extended/amended to cover the new DEFAULT COLLATION for table/view objects.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by dejankrak-db** [comment]
Adding @cloud-fan, @stefankandic, @stevomitric, @uros-db - please take a look when you find some time, thanks!

**Comment by dejankrak-db** [comment]
@stefankandic, @stevomitric, please take a look at the PR as discussed and let me know if you see any issues, thanks!

**Review by stefankandic — APPROVED** [review]
LGTM!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala
Comment (uros-db) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala:1]: can we add at least some test cases for the UTF8 family of collations? UTF8_BINARY, UTF8_LCASE

also RTRIM versions...

Comment (dejankrak-db) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala:1]: Sure, makes sense - I've extended the newly added DDLParserSuite test cases for creating/replacing/altering table to cover multiple different collations, including UTF8 family, UNICODE family (including RTRIM), as well as a couple of SR collations. Thanks for the suggestion!

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/parser/DataTypeAstBuilder.scala [hunk 1]
```diff
@@ -232,6 +232,16 @@ class DataTypeAstBuilder extends SqlBaseParserBaseVisitor[AnyRef] {
     visitMultipartIdentifier(ctx.collationName)
   }
 
+  protected def visitCollationSpecList(
+      ctx: java.util.List[CollationSpecContext]): Option[String] = {
+    ctx.asScala.headOption.map(visitCollationSpec)
+  }
+
+  override def visitCollationSpec(ctx: CollationSpecContext): String = withOrigin(ctx) {
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/parser/DataTypeAstBuilder.scala:240]: How are these two methods related to `DataTypeAstBuilder`? They are for Spark DDL commands and should be put in `AstBuilder` I think, the same as other CREATE TABLE/VIEW clauses.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 2]
```diff
@@ -119,6 +119,13 @@ class ResolveSessionCatalog(val catalogManager: CatalogManager)
         clusterBySpecOpt.getOrElse(ClusterBySpec(Nil)), conf.resolver))
       AlterTableSetPropertiesCommand(table.catalogTable.identifier, prop, isView = false)
 
+    case AlterTableCollation(ResolvedTable(catalog, _, table: V1Table, _), collation)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:122]: Since this is a new feature, I think we can directly go with the v2 command. i.e. do not match this command here, and eventually Spark will plan it as `AlterTableExec`

**Comment by cloud-fan** [comment]
To confirm: this PR is adding/extending the related DDL commands for specifying table/view level collation, and we will open a followup PR to leverage the table/view collation to determine default collations for input queries of DML commands?

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/parser/DataTypeAstBuilder.scala (hunk 1 above)
Comment (dejankrak-db) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/parser/DataTypeAstBuilder.scala:240]: Makes sense, moved these methods to AstBuilder.

**Comment by dejankrak-db** [comment]
> To confirm: this PR is adding/extending the related DDL commands for specifying table/view level collation, and we will open a followup PR to leverage the table/view collation to determine default collations for input queries of DML commands later?

@cloud-fan, that is correct, I've also updated the PR description with this info to avoid any confusion, thanks for pointing this out! I have addressed all your comments so far and refactored the code accordingly, please let me know if you have any further comments/concerns, or the PR can be merged, thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala (hunk 2 above)
Comment (dejankrak-db) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala]: @cloud-fan, actually this exposed a bug in v2AlterTableCommands implementation of AlterTableCollation(), which was not setting the collation properly for existing tables. I have changed this code to use setProperty instead (similar to how other table catalog props are altered), and added a test case in v2/DescribeTableSuite to ensure that this path now works correctly.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 3]
```diff
@@ -304,12 +304,13 @@ case class AlterTableSetPropertiesCommand(
...
-    // with the newly added/modified comment since CatalogTable also holds comment as its
-    // direct property.
+    // This overrides old properties and updates the comment and collation parameters of
+    // CatalogTable with the newly added/modified comment and collation since CatalogTable
+    // also holds comment and collation as its direct properties.
     val newTable = table.copy(
       properties = table.properties ++ properties,
-      comment = properties.get(TableCatalog.PROP_COMMENT).orElse(table.comment))
+      comment = properties.get(TableCatalog.PROP_COMMENT).orElse(table.comment),
+      collation = properties.get(TableCatalog.PROP_COLLATION).orElse(table.collation))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala:313]: how can we hit this code path? People use SET TBLPROPERTIES command and explicitly set the `collation` property?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/DescribeTableSuite.scala [hunk 4]
```diff
@@ -213,4 +213,29 @@ class DescribeTableSuite extends command.DescribeTableSuiteBase
       assert("""\d+\s+bytes,\s+4\s+rows""".r.matches(stats))
     }
   }
+
+  Seq(true, false).foreach { hasCollations =>
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/DescribeTableSuite.scala:217]: what's the difference between this one and the one in `v1/DescribeTableSuite`? Can we move it up to the parent suite?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala [hunk 5]
```diff
@@ -316,6 +335,17 @@ class HiveClientSuite(version: String) extends HiveVersionSuite(version) {
     assert(!client.tableExists("temporary", "tgt"))
   }
 
+  test("alterTable - change collation") {
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala:338]: is it the same with the test `create/alter table with collations`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala (hunk 3 above)
Comment (dejankrak-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala:313]: Actually, @cloud-fan, you are right, because we raise an error in AstBuilder.cleanTableProperties() when someone attempts to change collation through SET/UNSET TBLPROPERTIES, so no changes to the collation should be possible here or below in AlterTableUnsetPropertiesCommand - hence, I've reverted the changes in this file. Good catch!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala (hunk 5 above)
Comment (dejankrak-db) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala:338]: Right, this test case covers a subset of that one (just the alter table default collation part), so I have removed this test case.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/DescribeTableSuite.scala (hunk 4 above)
Comment (dejankrak-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/DescribeTableSuite.scala]: Moved the tests to DescribeTableSuiteBase suite, to avoid code duplication, whereas these test cases are now run as part of both v1 and v2 suites for complete coverage.

**Comment by dejankrak-db** [comment]
@cloud-fan, thanks for the review and great suggestions! I have addressed them all and changed/refactored the code accordingly, I believe the PR should be ready for merge, please let me know if you have any additional comments, thanks a lot!

**Comment by dejankrak-db** [comment]
Created https://issues.apache.org/jira/browse/SPARK-50522 and linked PR to it

**Comment by cloud-fan** [comment]
thanks, merging to master!
